### ISSUE TITLE
Incorrect HHI: preg_match can return bool

### ISSUE THREAD

**Issue by thewavelength** [facebook/hhvm#6679]
``` php
public static function validatePattern(string $pattern): bool
{
    /* HH_FIXME[4118] preg_match is only allowed to return int in HHVM type checker, but in reality in returns FALSE if the given regex is wrong... we want to check exactly this */
    return preg_match(':'.$pattern.':', null) !== false;
}
```

produces (without `HH_FIXME`):

```
.local/Helpers/StringHelper.hh:26:10,53: This expression is always true (Typing[4118])
  /tmp/hh_server/hhi_3435385d/stdlib/builtins_preg.hhi:21:88,90: This is an int
  .local/Helpers/StringHelper.hh:26:49,53: This is a bool
```

Should be valid imho without using `HH_FIXME`.

**Comment by jwatzman** [comment]
The decision here was intentional, since we felt that `preg_match` should only be used with constant patterns, for which you should never get a `false` return (invalid pattern). And an `HH_FIXME` can be used for the rare case you need it.

Of course, nothing actually enforces the patterns are string literals. We should probably do that, or revisit this decision. cc @int3

**Comment by lexidor** [comment]
I am going over old issues on this repository, to see which ones apply to the current versions of hhvm.

The HHI for `\preg_match()` has been updated to document this _lie_.
```
/*
 * `preg_match` can actually return false if the regex fails to compile.
 * However, most code has no need to consider this possibility because their
 * regexes are known at compile time. If you are using a regex only known at
 * runtime, please handle the possibility of a false return value, and override
 * any spurious type errors with a HH_FIXME comment.
 *
 * This will eventually be fixed with more type inference magic.
 */
```

The type inference magic is also here in the form of the `HH\Lib\Regex` api, released in hhvm [3.28.2](https://hhvm.com/blog/2018/09/28/hhvm-3.28.2.html) and promoted to stable in hhvm [4.0.0](https://hhvm.com/blog/2019/02/11/hhvm-4.0.0.html).
